observationunits-call

reynoldtan opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- No need to include the issue number in the title :-) -->
Get a set of observation units
https://brapi.docs.apiary.io/#reference/phenotypes/observationunits/get-observationunits

## Metadata
<!--- If it fixes an open issue, please add the issue link below. -->

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I feel this PR is ready to be merged.
- [ ] This PR is dependent upon [PR #/ nothing]

Documentation:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

## Description
<!--- Describe your changes in detail -->
<!--- Why is this change required? What problem does it solve? -->

## Dependencies
<!-- If this code is dependent upon another module and/or PR, 
       state that here. -->
<!-- Include information about other modules/PRs needed for testing,
        which to enable/merge first, etc. -->

## Testing?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how
      your change affects other areas of the code, etc. -->
<!--- Reviewers will use this section to test the submission! -->
- [ ] I tested on a generic Tripal Site
- [x] I tested on a KnowPulse Clone
- [ ] This PR includes automated testing

```
Call parameters: observationVariableDbId, germplasmDbId, studyDbId, page and pageSize.
Example:

host/web-services/brapi/v1/observationunits
host/web-services/brapi/v1/observationunits?observationVariableDbId=123
host/web-services/brapi/v1/observationunits?observationVariableDbId=123&germplasmDbId=123
host/web-services/brapi/v1/observationunits?studyDbId=123
host/web-services/brapi/v1/observationunits?pageSize=10
host/web-services/brapi/v1/observationunits?pageSize=5&page=3
```

reynoldtan committed:
observationunits-call

laceysanderson reviewed (changes requested):
Currently see a PDO error when trying to access `web-services/brapi/v1/observationunits`

```
PDOException: SQLSTATE[42704]: Undefined object: 7 ERROR: could not find array type for data type text[]: SELECT 'n/a' AS blocknumber, 'n/a' AS entrynumber, 'n/a' AS entrytype, germplasm.stock_id AS germplasmdbid, germplasm.name AS germplasmname, 'n/a' AS observationlevel, 'n/a' AS observationlevels, pheno.observable_id AS observationunitdbid, 'n/a' AS observationunitname, 'n/a' AS observationunitxref, 'n/a' AS id, 'n/a' AS source, ARRAY_TO_JSON(ARRAY_AGG(ARRAY[collector.value, pheno.phenotype_id::varchar, year.value::varchar, trait.cvterm_id::varchar, trait.name, 'n/a', pheno.value::varchar])) AS observations, 'n/a' AS plantnumber, 'n/a' AS plotnumber, 'n/a' AS positioncoordinatex, 'n/a' AS positioncoordinatextype, 'n/a' AS positioncoordinatey, 'n/a' AS positioncoordinateytype, program.value AS programname, replicate.value AS replicate, study.project_id AS studydbid, CONCAT(year.value, ' ', location.value) AS studylocation, 'n/a' AS studylocationdbid, study.name AS studyname, ARRAY_TO_JSON(ARRAY['n/a', 'n/a']) AS treatments, COUNT(germplasm.stock_id) OVER() AS full_count FROM chado.phenotype AS pheno INNER JOIN chado.stock AS germplasm USING (stock_id) INNER JOIN chado.project AS study USING (project_id) INNER JOIN chado.projectprop AS program ON study.project_id = program.project_id AND program.type_id = 6365 INNER JOIN chado.cvterm AS trait ON pheno.attr_id = trait.cvterm_id INNER JOIN chado.phenotypeprop AS collector ON pheno.phenotype_id = collector.phenotype_id AND collector.type_id = 6511 INNER JOIN chado.phenotypeprop AS year ON pheno.phenotype_id = year.phenotype_id AND year.type_id = 3788 INNER JOIN chado.phenotypeprop AS location ON pheno.phenotype_id = location.phenotype_id AND location.type_id = 6351 INNER JOIN chado.phenotypeprop AS replicate ON pheno.phenotype_id = replicate.phenotype_id AND replicate.type_id = 6510 GROUP BY pheno.observable_id, germplasm.stock_id, germplasm.name, study.project_id, study.name, location.value, year.value, program.value, replicate.value ORDER BY year.value DESC LIMIT 100 OFFSET 0 ; Array ( ) in chado_query() (line 1790 of /home/www/brapi-test/sites/all/modules/contrib/tripal/tripal_chado/api/tripal_chado.query.api.inc).
```

reynoldtan committed:
Revised observationunits call

reynoldtan committed:
Revised observationunits call

Fixed array of objects in response fields: observations,
observationunitxref and treatments

reynoldtan commented:
Revised call and using sample data generated by support module, will produce the sample response below.
![Screen Shot 2020-09-18 at 12 59 23 PM (2)](https://user-images.githubusercontent.com/15472253/93635061-df0cb880-f9ae-11ea-86c0-233ef1f80af0.png)

**Please note to use recent update to core class.**

laceysanderson reviewed (approved):
Will test later for same reason as locations call.
